File: android/app/src/main/java/com/bazical/app/ui/home/HomeViewModel.kt
package com.bazical.app.ui.home

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.bazical.app.domain.model.Gender
import com.bazical.app.domain.model.User
import com.bazical.app.domain.usecase.CalculateBaziUseCase
import com.bazical.app.domain.usecase.SolarToLunarUseCase
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale
import javax.inject.Inject

@HiltViewModel
class HomeViewModel @Inject constructor(
    private val calculateBaziUseCase: CalculateBaziUseCase,
    private val solarToLunarUseCase: SolarToLunarUseCase
) : ViewModel() {

    private val _uiState = MutableStateFlow(HomeUiState())
    val uiState: StateFlow<HomeUiState> = _uiState.asStateFlow()

    private val dateFormat = SimpleDateFormat("yyyy-MM-dd", Locale.getDefault())

    fun updateName(name: String) {
        _uiState.update { it.copy(name = name) }
    }

    fun updateBirthday(timestamp: Long) {
        _uiState.update { it.copy(birthday = timestamp) }
        convertSolarToLunar(timestamp)
    }

    fun updateBirthdayType(type: String) {
        _uiState.update { it.copy(birthdayType = type) }
    }

    fun showDatePicker() {
        _uiState.update { it.copy(showDatePicker = true) }
    }

    fun hideDatePicker() {
        _uiState.update { it.copy(showDatePicker = false) }
    }

    private fun convertSolarToLunar(timestamp: Long) {
        viewModelScope.launch {
            val dateStr = dateFormat.format(Date(timestamp))
            val result = solarToLunarUseCase(dateStr)
            if (result.isSuccess) {
                _uiState.update { it.copy(lunarDate = result.getOrNull()) }
            }
        }
    }

    fun calculate() {
        val state = _uiState.value
        if (state.birthday == null) {
            _uiState.update { it.copy(error = "请选择出生日期") }
            return
        }

        viewModelScope.launch {
            _uiState.update { it.copy(loading = true, error = null) }

            val birthdayStr = dateFormat.format(Date(state.birthday))
            val user = User(
                userId = "",
                name = "用户", // 默认名字
                birthday = birthdayStr,
                hour = 12, // 默认午时
                minute = state.minute,
                gender = state.gender,
                birthdayType = state.birthdayType
            )

            val result = calculateBaziUseCase(user)
            result.fold(
                onSuccess = {
                    _uiState.update { it.copy(loading = false, calculated = true) }
                },
                onFailure = { e ->
                    _uiState.update { it.copy(loading = false, error = e.message ?: "计算失败") }
                }
            )
        }
    }
}

File: ios/BaziCalApp/Features/Home/HomeViewModel.swift
import Foundation
import SwiftUI

/// 首页 UI 状态
struct HomeUiState {
    var birthday: Date?
    var timeValue: String = ""
    var hour: Int = 0
    var birthdayType: String = "solar"
    var showDatePicker: Bool = false
    var isLoading: Bool = false
    var calculated: Bool = false
    var error: String?
}

/// 首页视图模型
@MainActor
class HomeViewModel: ObservableObject {
    @Published var uiState = HomeUiState()
    @Published var calculatedBazi: Bazi?

    private let apiService = APIService.shared
    private let userDataStore = UserDataStore.shared

    init() {
        // 从本地存储加载已有数据
        if let bazi = userDataStore.userBazi {
            calculatedBazi = bazi
        }
    }

    /// 更新生日
    func updateBirthday(_ date: Date) {
        uiState.birthday = date
    }

    /// 更新时辰
    func updateTime(_ timeValue: String) {
        uiState.timeValue = timeValue
        uiState.hour = TimeLabel(rawValue: timeValue)?.hour ?? 0
    }

    /// 更新生日类型
    func updateBirthdayType(_ type: String) {
        uiState.birthdayType = type
    }

    /// 显示日期选择器
    func showDatePicker() {
        uiState.showDatePicker = true
    }

    /// 隐藏日期选择器
    func hideDatePicker() {
        uiState.showDatePicker = false
    }

    /// 计算八字
    func calculate() async {
        guard let birthday = uiState.birthday else {
            uiState.error = "请选择出生日期"
            return
        }

        uiState.isLoading = true
        uiState.error = nil

        do {
            let bazi = try await apiService.calculateBazi(
                birthday: birthday,
                hour: uiState.hour,
                minute: 0,
                birthdayType: BirthdayType(rawValue: uiState.birthdayType) ?? .solar
            )

            // 保存到本地
            userDataStore.userBazi = bazi
            userDataStore.hasCompletedOnboarding = true
            calculatedBazi = bazi
            uiState.isLoading = false
            uiState.calculated = true
        } catch {
            uiState.isLoading = false
            uiState.error = error.localizedDescription
        }
    }
}